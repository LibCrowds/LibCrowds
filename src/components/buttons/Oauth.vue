<template>
  <div class="oauth-buttons">

    <div v-if="facebook">
      <b-btn
        class="mx-1"
        variant="facebook"
        @click="redirect('facebook')">
        <icon name="facebook"></icon>
        <span class="ml-1 d-none d-sm-block">Facebook</span>
      </b-btn>
    </div>

    <div v-if="google">
      <b-btn
        class="mx-1"
        variant="googleplus"
        @click="redirect('google')">
        <icon name="google"></icon>
        <span class="ml-1 d-none d-sm-block">Google</span>
      </b-btn>
    </div>

    <div v-if="twitter">
      <b-btn
        class="mx-1"
        variant="twitter"
        @click="redirect('twitter')">
        <icon name="twitter"></icon>
        <span class="ml-1 d-none d-sm-block">Twitter</span>
      </b-btn>
    </div>

  </div>
</template>

<script>
import siteConfig from '@/siteConfig'

export default {
  props: {
    facebook: {
      type: Boolean,
      default: false
    },
    google: {
      type: Boolean,
      default: false
    },
    twitter: {
      type: Boolean,
      default: false
    }
  },

  methods: {
    /**
     * Redirect to an external endpoint for OAuth signin.
     * @param {String} endpoint
     *   The endpoint.
     */
    redirect (endpoint) {
      const nextExternal = `${document.location.origin}/${this.next}`
      const query = `next=${nextExternal}`
      window.location = `${siteConfig.pybossaHost}/${endpoint}?${query}`
    }
  }
}
</script>


<style lang="scss" scoped>
@import 'src/assets/style/main';

#oauth-buttons {
  display: flex;
  flex-direction: row;
  justify-content: center;

  .btn {
    @include hover-focus;
    border-radius: 0;
    color: $white;
    height: 100%;
    display: flex;

    @include hover-focus {
      color: $white;
    }

    &.btn-twitter {
      background: $twitter;

      @include hover-focus {
        background: darken($twitter, 5%);
      }
    }

    &.btn-facebook {
      background: $facebook;

      @include hover-focus {
        background: darken($facebook, 5%);
      }
    }

    &.btn-googleplus {
      background: $googleplus;

      @include hover-focus {
        background: darken($googleplus, 5%);
      }
    }
  }

  /* Fix inconsistend heights */
  svg {
    display: flex;
    height: 1.2rem;
  }
}
</style>
