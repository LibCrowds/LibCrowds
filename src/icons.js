import Vue from 'vue'
import Icon from 'vue-awesome/components/Icon'
import users from 'ionicons/dist/svg/ios-people-outline.svg'
import projects from 'ionicons/dist/svg/ios-desktop-outline.svg'
import taskruns from 'ionicons/dist/svg/ios-create-outline.svg'

Vue.component('icon', Icon)

Icon.register({
  'users': {
    width: 512,
    height: 512,
    raw: users
  }
})

Icon.register({
  'taskruns': {
    width: 512,
    height: 512,
    raw: taskruns
  }
})

Icon.register({
  'projects': {
    width: 512,
    height: 512,
    raw: projects
  }
})
