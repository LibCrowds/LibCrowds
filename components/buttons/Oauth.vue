<template>
  <div id="oauth-buttons">

    <b-btn
      v-if="facebook"
      variant="facebook"
      @click="redirect('facebook')">
      <icon name="facebook"></icon>
      Facebook
    </b-btn>

    <b-btn
      v-if="google"
      variant="googleplus"
      @click="redirect('google')">
      <icon name="google"></icon>
      Google
    </b-btn>

    <b-btn
      v-if="twitter"
      variant="twitter"
      @click="redirect('twitter')">
      <icon name="twitter"></icon>
      Twitter
    </b-btn>

  </div>
</template>

<script>
import 'vue-awesome/icons/facebook'
import 'vue-awesome/icons/google'
import 'vue-awesome/icons/twitter'
import localConfig from '@/local.config'

export default {
  props: {
    facebook: {
      type: Boolean,
      default: false
    },
    google: {
      type: Boolean,
      default: false
    },
    twitter: {
      type: Boolean,
      default: false
    },
    next: {
      type: String,
      required: ''
    }
  },

  methods: {
    /**
     * Redirect to an external endpoint for OAuth signin.
     * @param {String} endpoint
     *   The endpoint.
     */
    redirect (endpoint) {
      const query = `next=${this.next}&response_format=json`
      const url = `${localConfig.pybossaHost}/${endpoint}?${query}`
      window.location.assign(url)
    }
  }
}
</script>


<style lang="scss" scoped>
@import '~assets/style/settings';

#oauth-buttons {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  @include media-breakpoint-up (sm) {
    flex-direction: row;
  }

  .btn {
    @include hover-focus;
    display: flex;
    flex-direction: row;
    justify-content: center;
    width: 100%;
    color: $white;
    margin: 0.25rem 0.5rem;

    @include hover-focus {
      color: $white;
    }

    @include media-breakpoint-up (sm) {
      width: auto;
    }

    &.btn-twitter {
      background: $twitter;

      @include hover-focus {
        background: darken($twitter, 5%);
      }
    }

    &.btn-facebook {
      background: $facebook;

      @include hover-focus {
        background: darken($facebook, 5%);
      }
    }

    &.btn-googleplus {
      background: $googleplus;

      @include hover-focus {
        background: darken($googleplus, 5%);
      }
    }
  }

  /* Fix inconsistend heights */
  svg {
    display: flex;
    height: 1.2rem;
    margin-right: 0.5rem;
  }
}
</style>
