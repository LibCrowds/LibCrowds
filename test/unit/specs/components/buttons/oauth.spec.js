import BootstrapVue from 'bootstrap-vue'
import Icon from 'vue-awesome/components/Icon'
import 'vue-awesome/icons'
import 'vue-awesome/icons/facebook'
import 'vue-awesome/icons/google-plus'
import 'vue-awesome/icons/twitter'

import testLocalConfig from '~/test/test.local.config'
import { mount, createLocalVue } from 'vue-test-utils'
import Oauth from '~/components/buttons/Oauth'

describe('Oauth buttons', () => {
  const localVue = createLocalVue()
  localVue.use(BootstrapVue)
  localVue.component('icon', Icon)
  const next = '/'
  const wrapper = mount(Oauth, {
    localVue,
    propsData: {
      facebook: true,
      google: true,
      twitter: true,
      next: next
    }
  })

  it('render correctly', () => {
    const renderer = require('vue-server-renderer').createRenderer()
    renderer.renderToString(wrapper.vm, (err, str) => {
      expect(str).toMatchSnapshot()
    })
  })

  it('redirect to the correct oauth endpoint', () => {
    const endpoint = 'twitter'
    const host = testLocalConfig.pybossa.host
    const expectedUrl = `${host}/${endpoint}?next=${next}`
    const mockAssign = jest.fn()
    window.location.assign = mockAssign
    wrapper.vm.redirect(endpoint)
    expect(mockAssign).toHaveBeenCalledWith(expectedUrl)
  })
})
