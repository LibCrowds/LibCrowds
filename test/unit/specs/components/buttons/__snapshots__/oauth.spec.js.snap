// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Oauth buttons render correctly 1`] = `

<div id="oauth-buttons"
     data-server-rendered="true"
>
  <button type="button"
          class="btn btn-facebook"
  >
    <svg version="1.1"
         role="presentation"
         width="9.142857142857142"
         height="16"
         viewbox="0 0 1024 1792"
         class="fa-icon"
    >
      <path d="M959 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759h-306v-759h-255v-296h255v-218q0-186 104-288.5t277-102.5q147 0 228 12z">
      </path>
      
      <!---->
    </svg>
    Facebook
  </button>
  <button type="button"
          class="btn btn-googleplus"
  >
    <svg version="1.1"
         role="presentation"
         width="13.714285714285714"
         height="16"
         viewbox="0 0 1536 1792"
         class="fa-icon"
    >
      <path d="M768 786h725q12 67 12 128 0 217-91 387.5t-259.5 266.5-386.5 96q-157 0-299-60.5t-245-163.5-163.5-245-60.5-299 60.5-299 163.5-245 245-163.5 299-60.5q300 0 515 201l-209 201q-123-119-306-119-129 0-238.5 65t-173.5 176.5-64 243.5 64 243.5 173.5 176.5 238.5 65q87 0 160-24t120-60 82-82 51.5-87 22.5-78h-436v-264z">
      </path>
      
      <!---->
    </svg>
    Google
  </button>
  <button type="button"
          class="btn btn-twitter"
  >
    <svg version="1.1"
         role="presentation"
         width="14.857142857142858"
         height="16"
         viewbox="0 0 1664 1792"
         class="fa-icon"
    >
      <path d="M1620 408q-67 98-162 167 1 14 1 42 0 130-38 259.5t-115.5 248.5-184.5 210.5-258 146-323 54.5q-271 0-496-145 35 4 78 4 225 0 401-138-105-2-188-64.5t-114-159.5q33 5 61 5 43 0 85-11-112-23-185.5-111.5t-73.5-205.5v-4q68 38 146 41-66-44-105-115t-39-154q0-88 44-163 121 149 294.5 238.5t371.5 99.5q-8-38-8-74 0-134 94.5-228.5t228.5-94.5q140 0 236 102 109-21 205-78-37 115-142 178 93-10 186-50z">
      </path>
      
      <!---->
    </svg>
    Twitter
  </button>
</div>

`;
